packages/contracts-periphery/contracts/foundry-tests/Optimist.t.sol

```diff
@@ -0,0 +1,371 @@
+     */
+    function test_optimist_baseURI() external {
+        // set baseURI
+        AttestationStation.AttestationData[]
+            memory attestationData = new AttestationStation.AttestationData[](1);
+        attestationData[0] = AttestationStation.AttestationData({
+            about: address(optimist),
+            key: bytes32("optimist.base-uri"),
+            val: bytes(base_uri)
+        });
+        vm.prank(alice_admin);
+        attestationStation.attest(attestationData);
+
+        // assert baseURI is set
+        assertEq(optimist.baseURI(), base_uri);
+    }
+
+    /**
+     * @dev The tokenURI should return the token uri
+     * for a minted token
+     */
+    function test_optimist_token_uri() external {
+        // whitelist bob
+        // attest baseURI
+        AttestationStation.AttestationData[]
+            memory attestationData = new AttestationStation.AttestationData[](2);
+        // we are using true but it can be any non null value
+        attestationData[0] = AttestationStation.AttestationData({
+            about: bob,
+            key: bytes32("optimist.can-mint"),
+            val: bytes("true")
+        });
+        // we are using true but it can be any non null value
+        attestationData[1] = AttestationStation.AttestationData({
+            about: address(optimist),
+            key: bytes32("optimist.base-uri"),
+            val: bytes(base_uri)
+        });
+        vm.prank(alice_admin);
+        attestationStation.attest(attestationData);
+
+        // mint an NFT
+        vm.prank(bob);
+        optimist.mint(bob);
+
+        // assert tokenURI is set
+        assertEq(optimist.baseURI(), base_uri);
+        assertEq(
+            optimist.tokenURI(256),
+            // solhint-disable-next-line max-line-length
+            "https://storageapi.fleek.co/6442819a1b05-bucket/optimist-nft/attributes/0x0000000000000000000000000000000000000100.json"
+        );
+    }
+
+    /**
+     * @dev Should return a boolean of if the address is whitelisted
+     */
+    function test_optimist_is_whitelisted() external {
+        // whitelist bob
+        AttestationStation.AttestationData[]
+            memory attestationData = new AttestationStation.AttestationData[](1);
+        // we are using true but it can be any non null value
+        attestationData[0] = AttestationStation.AttestationData({
+            about: bob,
+            key: bytes32("optimist.can-mint"),
+            val: bytes("true")
+        });
+        vm.prank(alice_admin);
+        attestationStation.attest(attestationData);
+
+        // assert bob is whitelisted
+        assertEq(optimist.isWhitelisted(bob), true);
+        // assert sally is not whitelisted
+        assertEq(optimist.isWhitelisted(sally), false);
+    }
+
+    /**
+     * @dev Should return the token id of the owner
+     */
+    function test_optimist_token_id_of_owner() external {
+        // whitelist bob
+        AttestationStation.AttestationData[]
```

Comment:
There is a lot of repeated code:
- build attestation data array
- populate array
- prank as alice
- `CALL` attest

Perhaps modularize that into an `internal` function defined at the top of the contract that accepts the bare min args to build the array and make the attestation call

Reply:
Will do this in a follow up pr by end of week